project/listener.go

```diff
@@ -50,32 +25,93 @@ func NewDefaultListener(p *Project) chan<- events.Event {
 
 func (d *defaultListener) start() {
 	for event := range d.listenChan {
-		buffer := bytes.NewBuffer(nil)
-		if event.Data != nil {
-			for k, v := range event.Data {
-				if buffer.Len() > 0 {
-					buffer.WriteString(", ")
-				}
-				buffer.WriteString(k)
-				buffer.WriteString("=")
-				buffer.WriteString(v)
-			}
-		}
+		data, err := json.Marshal(event)
 
-		if event.EventType == events.ServiceUp {
+		switch event.(type) {
+		case *events.ServiceUpDone:
 			d.upCount++
 		}
 
 		logf := logrus.Debugf
 
-		if infoEvents[event.EventType] {
+		if infoLevel(event) {
 			logf = logrus.Infof
 		}
 
-		if event.ServiceName == "" {
-			logf("Project [%s]: %s %s", d.project.Name, event.EventType, buffer.Bytes())
+		if err != nil {
+			logf("Failed to Marshal Event [%s] for Project %s. Error: [%s]", event.String(), d.project.Name, err.Error())
+
+		} else if event.Service() == "" {
+			logf("Project [%s]: %s %s", d.project.Name, event.String(), data)
 		} else {
-			logf("[%d/%d] [%s]: %s %s", d.upCount, d.project.ServiceConfigs.Len(), event.ServiceName, event.EventType, buffer.Bytes())
+			logf("[%d/%d] [%s]: %s %s", d.upCount, d.project.ServiceConfigs.Len(), event.Service(), event.String(), data)
 		}
 	}
 }
+
+func infoLevel(event events.Event) bool {
```

Comment:
Is this something you could build in to the type? Or is it dependent on the scope of receiving the message? e.g. an `*events.ServiceDeleteStart` is only info level in the project listener, but debug elsewhere? If log level is responsible to the event, perhaps then you could build level in to the events type. e.g. `event.LogLevel()` ? Just a thought :)